### ISSUE TITLE
Hash code stub transformer implementation is incorrect

### ISSUE THREAD

**Issue by kkarnauk** [JetBrains/lincheck#131]
Version: 2.16 (and the previous ones, too)

**Issue**

As you can see in the implementation of [HashCodeStubTransformer](https://github.com/Kotlin/kotlinx-lincheck/blob/e27dff8ed99824be1c6da626ccafc4893c3d77bf/src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategyTransformer.kt#L618), in order to understand whether it should change a call of `hashCode`, it checkes `owner`. But it doesn't take into account that `owner` means the actual callee on the stack (not the concrete class which method will be called). 

So, there are following problems with that:

1. If we cast an object to `Any/Object` with overriden `hashCode` and call it, then `0` is returned. For example: 
```
val a: Any = 1
val b = 1
println(a.hashCode()) // prints 0
println(b.hashCode()) // prints 1
```  
I suppose, it is a huge hole. First of all, it becomes inconistent and the `equals-hashCode` invariant doesn't work. Also, in most collections all objects are casted to `Any/Object`, so all `hashCode` invocations return `0` and collections start working slow. 

2. If we have a class that doesn't override `hashCode` (and so do its parents) and `owner != Object`, then no transformation is done. It still calls the native implementation (which usually returns the memory address) and we have the non-determinism. 

**How to fix**

I have the following proposal. There are two fixes (one for each problem):

1. We check that `owner` is `Any/Object` and if it's true, we check `::class` of the value on the stack. If it is actually `Object/Any`, then we need to do the transformation.
2. We need another class transformer. It should check if there is a class which inherits `Object` directly and doesn't override the default `hashCode()`, it should add an implementation of `hashCode` which returns 0.

**Comment by ndkoval** [comment]
To solve the issue, we need to track down the classes that override hashCode. One unreliable but simple solution would be to check whether `obj.hashCode() == System.identityHashCode(obj)` to detect that `hashCode()` returns the identity hash code. Thanks to @kkarnauk for the suggestion. While this simple solution is imperfect, it is super-easy to implement and should work fine in practice.

**Comment by ndkoval** [comment]
Should be resolved in #249
